core/lib/zksync_core/src/sync_layer/snapshots.rs

```diff
@@ -0,0 +1,496 @@
+    ) -> anyhow::Result<SnapshotApplier<'a, 'b, 'd>, SnapshotApplierError> {
+        let mut storage = connection_pool
+            .access_storage_tagged("snapshots_applier")
+            .await
+            .unwrap();
+        let mut storage = storage.start_transaction().await.unwrap();
+
+        let mut applied_snapshot_status = storage
+            .applied_snapshot_status_dal()
+            .get_applied_snapshot_status()
+            .await
+            .unwrap();
+
+        if !storage.blocks_dal().is_genesis_needed().await.unwrap()
+            && applied_snapshot_status.is_none()
+        {
+            return Err(Canceled(
+                "This node has already been initialized without a snapshot".to_string(),
+            ));
+        }
+
+        if applied_snapshot_status.is_some()
+            && applied_snapshot_status.as_ref().unwrap().is_finished
+        {
+            return Err(Canceled(
+                "This node has already been initialized from a snapshot".to_string(),
+            ));
+        }
+
+        let snapshot_response = client.fetch_newest_snapshot().await.unwrap();
+        if snapshot_response.is_none() {
+            return Err(Canceled("Main node does not have any ready snapshots, skipping initialization from snapshot!".to_string()));
+        }
+        let snapshot = snapshot_response.unwrap();
+        let l1_batch_number = snapshot.l1_batch_number;
+        tracing::info!("Found snapshot with data up to l1_batch {}, created at {}, storage_logs are divided into {} chunk(s)", l1_batch_number, snapshot.generated_at, snapshot.storage_logs_chunks.len());
+
+        if applied_snapshot_status.is_none() {
+            applied_snapshot_status = Some(AppliedSnapshotStatus {
+                l1_batch_number,
+                is_finished: false,
+                last_finished_chunk_id: None,
+            });
+        };
+        let applied_snapshot_status = applied_snapshot_status.unwrap();
+
+        let recovered_version = l1_batch_number.0 as u64;
+
+        let rocks_db = RocksDBWrapper::new(Path::new(&merkle_tree_db_path));
+
+        let recovery = MerkleTreeRecovery::new(rocks_db, recovered_version);
+
+        let blob_store = ObjectStoreFactory::snapshots_from_env()
+            .context("ObjectStoreFactor::snapshots_from_env()")?
+            .create_store()
+            .await;
+
+        storage
+            .applied_snapshot_status_dal()
+            .set_applied_snapshot_status(&applied_snapshot_status)
+            .await
+            .unwrap();
+        Ok(Self {
+            connection_pool,
+            client,
+            recovery,
+            blob_store,
+            applied_snapshot_status,
+            snapshot,
+            state_keeper_config,
+            metadata_calculator_config,
+        })
+    }
+
+    async fn build_state_keeper(config: StateKeeperConfig) -> ZkSyncStateKeeper {
+        let (_, stop_receiver) = watch::channel(false);
```

Comment:
I don't understand how this is supposed to work. Since the stop sender is immediately dropped, the components initialized with `stop_receiver` should immediately stop. Is this their intended behavior? Why are these components (e.g., fetcher) are started and then waited for anyway? Components are supposed to be spawned as Tokio tasks and run until the program termination; i.e., if a component is waited for, it is a logical problem on its own.

Reply:
By adding stop_after_one_batch my intention was to modify those components to allow running them only until certain point in L1 history. This way, after the restore process there exists exactly one L1 batch in db that looks exactly like it would if it was inserted during normal state_keeper operation. We also don't need to worry if snapshot applier for instance 'forgot' to populate one table. Most changes in the codebase won't have to take into consideration that the snapshot applier exists. I suggest a huddle tomorrow to talk more about this change